### PR TITLE
Add a `return_outputs` flag to `predict()`

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
This PR adds a `return_outputs` flag to `predict()`, which when set, will move all batch outputs to cpu and accumulate them in a python list when running `predict()`.

Closes https://mosaicml.atlassian.net/browse/CO-765

### CODE REVIEW

**Review by mvpatel2000 — APPROVED** [review]
LGTM

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -1680,7 +1680,7 @@ def _train_batch(self, use_grad_scaling: bool):
                     'least one rank encountered a CUDA OOM error.') from e
             # Synchronize new batch compute time
             batch_compute_time = end_time - start_time
-            batch_compute_time = self._device.tensor_to_device(torch.tensor([batch_compute_time], dtype=torch.uint8))
+            batch_compute_time = self._device.tensor_to_device(torch.tensor([batch_compute_time], dtype=torch.float))
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1683]: good catch

File: composer/trainer/trainer.py [hunk 2]
```diff
@@ -1884,6 +1946,9 @@ def predict(self, dataloader: Union[DataLoader, DataSpec], subset_num_batches: i
                 self.state.outputs = self.state.model(self.state.batch)
                 self.engine.run_event(Event.PREDICT_AFTER_FORWARD)
 
+                if return_outputs:
+                    outputs.append(cpu_device.batch_to_device(self.state.outputs))
```
Comment (hanlint) [inline composer/trainer/trainer.py:1950]: should we concatenate along the batch dimension so its one tensor/np array?

File: composer/trainer/trainer.py [hunk 3]
```diff
@@ -1836,6 +1842,59 @@ def predict(self, dataloader: Union[DataLoader, DataSpec], subset_num_batches: i
...
+
+                    If collecting all outputs from the dataloader would exceed available memory, we recommend
+                    to set `return_outputs` to False and use your own callback to write the outputs from each
+                    prediction batch to a file. For example:
+
+                    .. testsetup::
+
+                        predict_dl = train_dataloader
+
+                    .. testcode::
```
Comment (hanlint) [inline composer/trainer/trainer.py:1859]: should we just make this its won callback and backlink it here?

Comment (hanlint) [reply composer/trainer/trainer.py:1859]: if not, we should move this into the main body of the docstrings as a "There are two ways to access the outputs of `predict`. With `return_outputs=True`, you can use the list of answers and save them to disk yourself, e.g. `outputs = trainer.predict(..); torch.save(outputs, ..)`.

Or if the total outputs are too large, use `return_outputs=False`, and add the predictionCallback to save things to disk or object store  with each batch.

File: composer/trainer/trainer.py (hunk 2 above)
Comment (eracah) [reply composer/trainer/trainer.py:1950]: +1

File: tests/trainer/test_predict.py [hunk 4]
```diff
@@ -28,6 +33,21 @@ def _assert_predict_events_called_expected_number_of_times(
         assert expected == actual, f'Event {event} expected to be called {expected} times, but instead it was called {actual} times'
 
 
+class PredictionSaver(Callback):
+
+    def __init__(self, folder: str):
+        self.folder = folder
+        os.makedirs(self.folder, exist_ok=True)
+
+    def predict_batch_end(self, state: State, logger: Logger) -> None:
```
Comment (eracah) [inline tests/trainer/test_predict.py:42]: comment: there might be a situation, where the logger has a `log_table` function that we call here

**Review by eracah — APPROVED** [review]

File: composer/trainer/trainer.py (hunk 2 above)
Comment (ravi-mosaicml) [reply composer/trainer/trainer.py:1950]: I was thinking about this.....and was leaning against it, as the outputs _could_ be different shapes. Or perhaps the output could be a dictionary. IDK, thoughts?

File: composer/trainer/trainer.py (hunk 3 above)
Comment (ravi-mosaicml) [reply composer/trainer/trainer.py:1859]: I'd rather leave it as a docstring for now until we can write a robust callback. But agreed, I'll put this high level in the docstring.

File: composer/trainer/trainer.py (hunk 2 above)
Comment (hanlint) [reply composer/trainer/trainer.py:1950]: Oh, good point... Don't we have smart infer for concat based on type, or is that just for split? I dont think this is a blocker for merging

Comment (eracah) [reply composer/trainer/trainer.py:1950]: what about using a nestedtensor for outputs of different shapes? https://pytorch.org/docs/stable/nested.html

Comment (hanlint) [reply composer/trainer/trainer.py:1950]: I'd hesitate to have our public API depends on something in prototype stage at this point.

File: composer/trainer/trainer.py (hunk 1 above)
Comment (ravi-mosaicml) [reply composer/trainer/trainer.py:1683]: Doctests were failing locally for me...

File: composer/trainer/trainer.py (hunk 2 above)
Comment (ravi-mosaicml) [reply composer/trainer/trainer.py:1950]: I'll leave this as-is for now. We can always add in a collation argument later to try to consolidate it into one data structure. The user can pass a custom collate fn as part of their data spec, but we don't have a generic collate fn like we do with a batch split one.

File: composer/trainer/trainer.py (hunk 3 above)
Comment (ravi-mosaicml) [reply composer/trainer/trainer.py]: Updated the docstring.

**Pull request merged**